dynamicio/io/s3.py

```diff
@@ -0,0 +1,82 @@
+from functools import partial
+from pathlib import Path
+from typing import Any, Dict, List, Literal, Optional, Type
+
+import boto3
+import pandas as pd
+
+from dynamicio.io.s3_contexts import s3_named_file_reader, s3_writer, s3_reader
+from dynamicio.io.resource import BaseResource
+from dynamicio.io.serde import BaseSerde, CsvSerde, HdfSerde, JsonSerde, ParquetSerde, PickleSerde
+
+
+class S3Resource(BaseResource):
+    bucket: str
+    path: Path
+    read_kwargs: Dict[str, Any] = {}
+    write_kwargs: Dict[str, Any] = {}
+    injectables: List[str] = ["path"]
+    file_type: Optional[Literal["parquet", "hdf", "csv", "json", "pickle"]] = None
+    force_read_to_memory: bool = False
+
+    @property
+    def _s3_path(self) -> str:
+        """For logging purposes only."""
+        return f"s3://{self.bucket}/{self.path}"
+
+    def _read(self) -> pd.DataFrame:
+        if self.force_read_to_memory:
+            with s3_reader(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as fobj:  # type: ignore
+                df = self.get_serde()._read(fobj, **self.read_kwargs)  # type: ignore
+                if df is not None:
+                    return df
+
+        with s3_named_file_reader(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as target_file:
+            return self.get_serde()._read(target_file.name, **self.read_kwargs)  # type: ignore
+
+    def _write(self, df: pd.DataFrame) -> None:
+        with s3_writer(boto3.client("s3"), s3_bucket=self.bucket, s3_key=str(self.path)) as fobj:
+            return self.get_serde()._write(fobj, df)
```

Comment:
nit: should we make the `_write` method in the Serdes public? the fact that we are using it as a private method here is a bit of a code smell

Reply:
Same thing for `_read`

Reply:
agreed.

Reply:
I have read again the `Serde` implementations. I see why we kept that private. I cannot think of a nice re-design from the top of my mind. It's a minor thing as well, so worth refactoring this later on

Reply:
The reason being that the main purpose of the serdes are read/write _from_file?
I think we can still make them public so we don't call _read _write in other places... Even though what it really boils down to is that we shouldn't be using the serde's within the resources like that. Anyway, it works for now.